File: src/conventions/core/pipeline/gate-order.ts
/** The 4-gate sequence that every proposal passes through. */
export type GateId =
  | "semantic-applicability"
  | "obligation-satisfaction"
  | "encoder-availability"
  | "concrete-legality";

export interface GateResult {
  readonly gateId: GateId;
  readonly passed: boolean;
  readonly reason?: string;
}

export interface GatedProposal {
  readonly candidateBidName: string;
  readonly moduleId: string;
  readonly gateResults: readonly GateResult[];
  readonly passedAllGates: boolean;
}

/** The canonical gate evaluation order. */
const GATE_ORDER: readonly GateId[] = [
  "semantic-applicability",
  "obligation-satisfaction",
  "encoder-availability",
  "concrete-legality",
] as const;

/** Evaluate a candidate against the gate sequence. Returns on first failure. */
export function evaluateGates(
  checks: readonly {
    gateId: GateId;
    passed: boolean;
    reason?: string;
  }[],
): { results: readonly GateResult[]; passedAll: boolean } {
  const results: GateResult[] = [];
  for (const gateId of GATE_ORDER) {
    const check = checks.find((c) => c.gateId === gateId);
    if (!check) {
      // Gate not evaluated — treat as passed (forward-compatible)
      results.push({ gateId, passed: true });
      continue;
    }
    results.push({ gateId, passed: check.passed, reason: check.reason });
    if (!check.passed) {
      return { results, passedAll: false };
    }
  }
  return { results, passedAll: true };
}

File: src-tauri/crates/bridge-conventions/src/pipeline/evaluation/gate_order.rs
//! Gate order — 4-gate sequence evaluation.
//!
//! Mirrors TS from `pipeline/evaluation/gate-order.ts`.

/// Gate check result.
pub struct GateResult {
    pub passed: bool,
    pub gate_id: String,
    pub reason: Option<String>,
}

/// 4-gate sequence:
/// 1. semantic-applicability — does the hand satisfy all clauses?
/// 2. obligation-satisfaction — does the bid satisfy any forcing obligation?
/// 3. encoder-availability — can the meaning be encoded as a legal call?
/// 4. concrete-legality — is the resolved call legal in the current auction?
pub fn evaluate_gates(
    hand_satisfied: bool,
    encoding_legal: bool,
) -> Vec<GateResult> {
    let mut results = Vec::new();

    // Gate 1: semantic-applicability
    results.push(GateResult {
        passed: hand_satisfied,
        gate_id: "semantic-applicability".into(),
        reason: if hand_satisfied {
            None
        } else {
            Some("Hand does not satisfy all clauses".into())
        },
    });

    if !hand_satisfied {
        return results;
    }

    // Gate 2: obligation-satisfaction (always passes for now — Phase 4)
    results.push(GateResult {
        passed: true,
        gate_id: "obligation-satisfaction".into(),
        reason: None,
    });

    // Gate 3: encoder-availability
    results.push(GateResult {
        passed: encoding_legal,
        gate_id: "encoder-availability".into(),
        reason: if encoding_legal {
            None
        } else {
            Some("No legal encoding available".into())
        },
    });

    if !encoding_legal {
        return results;
    }

    // Gate 4: concrete-legality
    results.push(GateResult {
        passed: encoding_legal,
        gate_id: "concrete-legality".into(),
        reason: None,
    });

    results
}
